# MatSE580 Guest Lecture 2
## Introduction

In this guest lecture we will cover:
1. [Interacting with the database we set up in Lecture 1](#verify-the-connection-to-the-database) and [visualizing the results](#plotting-with-mongodb-charts) - using [pymongo](https://github.com/mongodb/mongo-python-driver) library and [MongoDB Charts](https://www.mongodb.com/docs/charts/) service
2. [Using machine learning (ML) tools to predict stability of materials](#pysipfenn) - using [pySIPFENN](https://pysipfenn.readthedocs.io/en/stable/)
3. [Using ML featurization and dimensionality reduction to embed materials in feature space](#featurization) - using [pySIPFENN](https://pysipfenn.readthedocs.io/en/stable/) with [MongoDB Charts](https://www.mongodb.com/docs/charts/) visualization
4. [Using faturization to guide DFT and improve ML models](#transfer-learning-on-small-dft-dataset)

**This notebook assumes that you already followed the instructions in Lecture 1 and you:**
1. Have a conda environment called `580demo` (or other) with all the packages installed, including:
    - `pymatgen`
    - `pymongo`
    - `pysipfenn`

2. Have a MongoDB database called `matse580` with collection `structures` to which you have access:
    - username (e.g. `student`)
    - API key / password string (e.g. `sk39mIM2f35Iwc`)
    - whitelisted your IP address or `0.0.0.0/0` (entire internet)
    - know the connection string (URI) to the database (e.g. `mongodb+srv://student:sk39mIM2f35Iwc@cluster0.3wlhaan.mongodb.net/?retryWrites=true&w=majority`)

3. You populated the database with all Sigma phase end members (see Lecture 1 - Inserting Data)

4. After you installed `pysipfenn`, you have downloaded all the [pre-trained models](https://zenodo.org/records/7373089) by calling `downloadModels()` and it finished successfully. If not, run this one liner:

        python -c "import pysipfenn; c = pysipfenn.Calculator(); c.downloadModels(); c.loadModels();"

If all of the above are true, you are ready to go!

In [230]:
from pprint import pprint            # pretty printing
from collections import defaultdict  # convenience in the example
import os                            # file handling
from datetime import datetime        # time handling
from zoneinfo import ZoneInfo        # time handling
from pymatgen.core import Structure  # pymatgen
import numpy as np                   # numpy for data manipulation
import plotly.express as px          # plotly for plotting
from importlib import resources      # for accessing the data files

## Verify the connection to the database
pymongo is a Python library that allows us to interact with MongoDB databases in a very intuitive way. Let's start by importing its `MongoClient` class and creating a connection to our database:

In [2]:
from pymongo import MongoClient
uri = 'mongodb+srv://amk7137:kASMuF5au1069Go8@cluster0.3wlhaan.mongodb.net/?retryWrites=true&w=majority'
client = MongoClient(uri)

and see what databases are available:

In [3]:
client.list_database_names()

['matse580', 'admin', 'local']

Now connect to `matse580\structures` collection

In [4]:
collection = client['matse580']['structures']

and verify that the Sigma phase structures we created are there:

In [5]:
print(f'Found: {collection.count_documents({})} structures\n')
pprint(collection.find_one({}, skip=100))

Found: 243 structures

{'POSCAR': 'Cr12 Fe10 Ni8\n'
           '1.0\n'
           '   8.5470480000000002    0.0000000000000000    0.0000000000000000\n'
           '   0.0000000000000000    8.5470480000000002    0.0000000000000000\n'
           '   0.0000000000000000    0.0000000000000000    4.4777139999999997\n'
           'Cr Fe Ni Fe Cr\n'
           '8 2 8 8 4\n'
           'direct\n'
           '   0.7377020000000000    0.0637090000000000    0.0000000000000000 '
           'Cr\n'
           '   0.2622980000000000    0.9362910000000000    0.0000000000000000 '
           'Cr\n'
           '   0.4362910000000000    0.2377020000000000    0.5000000000000000 '
           'Cr\n'
           '   0.7622980000000000    0.5637090000000000    0.5000000000000000 '
           'Cr\n'
           '   0.5637090000000000    0.7622980000000000    0.5000000000000000 '
           'Cr\n'
           '   0.2377020000000000    0.4362910000000000    0.5000000000000000 '
           'Cr\n'
           '   0.0637

### Plotting with MongoDB Charts

MongoBD Charts is an associated service that allows us to quickly visualize the data in the database online and share it with others, while keeping the source data secure and private.

***Note for Online Students: At this point we will pause the Jupiter Notebook and switch to the MongoDB Atlas website to set up the database, or skip until next week depending on the available time.** The process is fairly straightforward but feel free to stop by office hours for help!*

You should end up with some neat figures like the one below 

<p align="center">
  <img src="assets/MongoDBChartExample.png" width="500"/>
</p>

If you are interested in seeing a couple more examples, you can visit the dashboard of [ULTERA Database](https://ultera.org) for high entropy alloys.

## pySIPFENN

We will now complete a brief walkthrough covering core functionalities of the **pySIPFENN** or **py**(**S**tructure-**I**nformed **P**rediction of **F**ormation **E**nergy using **N**eural **N**etworks) package; available through the PyPI repository. For a full up-to-date documentation, please refer to the [pySIPFENN documentation page](https://pysipfenn.org) or [pySIPFENN GitHub repository](https://git.pysipfenn.org). You can also find news about our projects using SIPFENN at our [Phases Research Lab](https://phaseslab.org) group website.

On the conceptual level, pySIPFENN is a framework composed of:

- Featurizers / descriptor calculators allowing user to interpret atomic structures (hence **S**tructure-**I**nformed) and represent them with numbers in a way suitable for machine learning (ML) **P**rediction of properties. A few shipped to public are Ward2017 (general) and KS2022 (general or optimized to different material types) calcualting Ward2017 and KS2022 feature vectors, respectively. Thanks to how modular pySIPFENN is, you can generally just "steal" them as standalone modules and use them in your own projects.

- It can handle any properties user wants to predict based purely on the model training, but the key fundamental property of interest has been **F**ormation **E**nergy of materials and that is what is shipped by default with the package.

- It can use any [Open Neural Network Exchange (ONNX)](https://onnx.ai) trained on the supported feature vectors (Ward2017 and KS2022 included). The models shipped by default are **N**eural **N**etworks, hence inclusion in the name, but neither pySIPFENN nor ONNX is limited to NNs. You can export, for instance complete `scikit-learn` pipelines (as done [here in heaGAN package](https://github.com/amkrajewski/cGAN_demo/blob/master/heagan/notebooks/train_surrogates.ipynb)) and use them in pySIPFENN.

The figure below shows how they fit together conceptually:

<p align="center">
  <img src="assets/neuralnetcolorized.png" width="500"/>
</p>

### Getting Started

To utilize pySIPFENN for straightforward calculations, **only the Calculator class is needed**, which acts as an ***environment*** for all components of the package. Under the hood, it will do a lot of things for you, including both fetching and identification of available NN models. Afterwards, it will expose a very high-level API for you to use. 

In [6]:
from pysipfenn import Calculator     # The only thing needed for calculations

Could not import coremltools.

Dependencies for exporting to CoreML, Torch, and ONNX are not installed by default with pySIPFENN. You need to install pySIPFENN in "dev" mode like: pip install -e "pysipfenn[dev]", or like pip install -e ".[dev]" ifyou are cloned it. See pysipfenn.org for more details.


Now initialize the Calculator. When run, this should display all models detected (e.g. ✔ SIPFENN_Krajewski2020 Standard Materials Model)
and those not detected, but declared in the `modelsSIPFENN/models.json` file. If some networks are not detected (prepended with *x*), this may mean download (you were to do in Lecture 1) was not completed successfully. You can try to download them again by calling `c.downloadModels()` which will only download the missing ones.

In [7]:
c = Calculator()

*********  Initializing pySIPFENN Calculator  **********
Loading model definitions from: /Users/adam/opt/anaconda3/envs/580demo/lib/python3.10/site-packages/pysipfenn/modelsSIPFENN/models.json
Found 4 network definitions in models.json
✔ SIPFENN_Krajewski2020 Standard Materials Model
✔ SIPFENN_Krajewski2020 Novel Materials Model
✔ SIPFENN_Krajewski2020 Light Model
✔ SIPFENN_Krajewski2022 KS2022 Novel Materials Model
Loading all available models (autoLoad=True)
Loading models:


  0%|          | 0/4 [00:00<?, ?it/s]

100%|██████████| 4/4 [00:14<00:00,  3.65s/it]

*********  pySIPFENN Successfully Initialized  **********


The simplest and most common usage of pySIPFENN is to deploy it on a directory/folder containing atomic structure files such as POSCAR or CIF. To of so, one simply specifies its location and which descriptor / feature vector should be used. The latter determines which ML models will be run, as they require a list of specific and ordered features as input.

    c.runFromDirectory(directory='myInputFiles', descriptor='KS2022')

Furthermore, while the exact model can be specified by the user, by default all applicable models are run, as the run itself is 1-3 orders of magnitude faster than descriptor calculation. Following the link printed during `Calculator` initialization reveals which models will be run.

In this demonstration, a set of test files shipped under `assets/examplePOSCARS`. Let's run them with Ward2017 featurizer.

In [8]:
c.runFromDirectory(directory='assets/examplePOSCARS',
                   descriptor='Ward2017');

Importing structures...


100%|██████████| 6/6 [00:00<00:00, 21.46it/s]



Models that will be run: ['SIPFENN_Krajewski2020_NN9', 'SIPFENN_Krajewski2020_NN20', 'SIPFENN_Krajewski2020_NN24']
Calculating descriptors...


100%|██████████| 6/6 [00:05<00:00,  1.20it/s]


Done!
Making predictions...
Prediction rate: 20.5 pred/s
Obtained 6 predictions from:  SIPFENN_Krajewski2020_NN9
Prediction rate: 21.5 pred/s
Obtained 6 predictions from:  SIPFENN_Krajewski2020_NN20
Prediction rate: 131.0 pred/s
Obtained 6 predictions from:  SIPFENN_Krajewski2020_NN24
Done!


Now, all results are obtained and stored within the **c** Calculator object inside a few exposed conveniently named variables
_predictions_ and _inputFiles_. Also, the descriptor data is retained in _descriptorData_ if needed. Let's look up all 6 entries. Note that the unit of prediction will depend on the model used; in this case, it is eV/atom.

In [9]:
pprint(c.inputFiles)
pprint(c.predictions)

['12-Gd4Cr4O12.POSCAR',
 '13-Fe16Ni14.POSCAR',
 '14-Fe24Ni6.POSCAR',
 '15-Ta4Tl4O12.POSCAR',
 '16-Fe18Ni12.POSCAR',
 '17-Pr4Ga4O12.POSCAR']
[[-3.154766321182251, -3.214848756790161, -3.187128782272339],
 [-0.013867354951798916, 0.04655897989869118, 0.053411152213811874],
 [0.02639671415090561, 0.05997598543763161, 0.06677809357643127],
 [-2.467507839202881, -2.4308743476867676, -2.391871690750122],
 [0.01810809224843979, 0.06462040543556213, 0.10881152749061584],
 [-2.7106518745422363, -2.6583476066589355, -2.727781057357788]]


For user convenience, a few methods are provided for extracting the results. E.g., if pySIPFENN has been run from structure files, the `get_resultDictsWithNames()` method is available to conveniently pass results forward in the code.

In [10]:
c.get_resultDictsWithNames()

[{'name': '12-Gd4Cr4O12.POSCAR',
  'SIPFENN_Krajewski2020_NN9': -3.154766321182251,
  'SIPFENN_Krajewski2020_NN20': -3.214848756790161,
  'SIPFENN_Krajewski2020_NN24': -3.187128782272339},
 {'name': '13-Fe16Ni14.POSCAR',
  'SIPFENN_Krajewski2020_NN9': -0.013867354951798916,
  'SIPFENN_Krajewski2020_NN20': 0.04655897989869118,
  'SIPFENN_Krajewski2020_NN24': 0.053411152213811874},
 {'name': '14-Fe24Ni6.POSCAR',
  'SIPFENN_Krajewski2020_NN9': 0.02639671415090561,
  'SIPFENN_Krajewski2020_NN20': 0.05997598543763161,
  'SIPFENN_Krajewski2020_NN24': 0.06677809357643127},
 {'name': '15-Ta4Tl4O12.POSCAR',
  'SIPFENN_Krajewski2020_NN9': -2.467507839202881,
  'SIPFENN_Krajewski2020_NN20': -2.4308743476867676,
  'SIPFENN_Krajewski2020_NN24': -2.391871690750122},
 {'name': '16-Fe18Ni12.POSCAR',
  'SIPFENN_Krajewski2020_NN9': 0.01810809224843979,
  'SIPFENN_Krajewski2020_NN20': 0.06462040543556213,
  'SIPFENN_Krajewski2020_NN24': 0.10881152749061584},
 {'name': '17-Pr4Ga4O12.POSCAR',
  'SIPFENN_Kr

Alternatively, if results are to be preserved in a spreadsheet, they can be exported into a CSV.

In [11]:
c.writeResultsToCSV('myFirstResults_pySIPFENN.csv')

### Predicting all Sigma Endmembers from Lecture 1

Now, armed with power of pySIPFENN, we can quickly get formation energies of all Sigma phase endmembers we defined in last lectrue. We start by getting all the structures from the database:

In [22]:
structList, idList = [], []
for entry in collection.find({}):
    idList.append(entry['_id'])
    structList.append(Structure.from_dict(entry['structure']))
print(f'Fetched {len(structList)} structures')

Fetched 243 structures


Now, we will use `runModels` function, which is one layer of abstraction lower than `runFromDirectory` as it skips file processing and directly takes the structure objects. We will set `mode='parallel'` to run in parallel, which is much faster than sequential execution on multi-core machines. Each thread on a modern CPU should be able to process ~1 structure per second, so this should take about a minute.

We will also use `get_resultDicts` to get the results in a convenient format. 

In [14]:
c.runModels(structList=structList, descriptor='Ward2017', mode='parallel', max_workers=4)
results = c.get_resultDicts()


Models that will be run: ['SIPFENN_Krajewski2020_NN9', 'SIPFENN_Krajewski2020_NN20', 'SIPFENN_Krajewski2020_NN24']
Calculating descriptors...


  0%|          | 0/243 [00:00<?, ?it/s]

Could not import coremltools.

Dependencies for exporting to CoreML, Torch, and ONNX are not installed by default with pySIPFENN. You need to install pySIPFENN in "dev" mode like: pip install -e "pysipfenn[dev]", or like pip install -e ".[dev]" ifyou are cloned it. See pysipfenn.org for more details.
Could not import coremltools.

Dependencies for exporting to CoreML, Torch, and ONNX are not installed by default with pySIPFENN. You need to install pySIPFENN in "dev" mode like: pip install -e "pysipfenn[dev]", or like pip install -e ".[dev]" ifyou are cloned it. See pysipfenn.org for more details.Could not import coremltools.
Could not import coremltools.


Dependencies for exporting to CoreML, Torch, and ONNX are not installed by default with pySIPFENN. You need to install pySIPFENN in "dev" mode like: pip install -e "pysipfenn[dev]", or like pip install -e ".[dev]" ifyou are cloned it. See pysipfenn.org for more details.

Dependencies for exporting to CoreML, Torch, and ONNX are not i

In [18]:
pprint(results[0])

{'SIPFENN_Krajewski2020_NN20': 0.07977379858493805,
 'SIPFENN_Krajewski2020_NN24': 0.03619053587317467,
 'SIPFENN_Krajewski2020_NN9': 0.07845475524663925}


and now we can easily upload them back to the database, as we learned in Lecture 1

In [23]:
for id, result in zip(idList, results):
    collection.update_one({'_id': id}, {'$set': result})

and now they are accessible to anyone with access!

In [25]:
collection.find_one({}, skip=100)['SIPFENN_Krajewski2020_NN9']

0.15312525629997253

## Featurization

We just made some predictions using pySIPFENN and created a dataset to share with the world, but like with most ML materials tools, the real power comes from featurization, which is often obfuscated from the user. Fortunately, pySIPFENN is very transparent and allows us to easily access all `Ward2017` features for all structures we just predicted under `descriptorData` variable of the `Calculator` object.

In [30]:
print(f'Number of features: {len(c.descriptorData[25])}\nFeature values:')
pprint(c.descriptorData[25])

Number of features: 271
Feature values:
array([1.26641789e+01, 5.69714967e-01, 1.17938896e+01, 1.37944770e+01,
       1.61693086e-02, 9.69056913e-01, 1.03318127e+00, 5.20789292e-02,
       1.48544654e-02, 2.72743042e-02, 6.61336996e-02, 3.46242699e-02,
       2.02959457e-01, 7.97080010e-02, 3.87759431e-02, 5.79071913e-01,
       1.58050955e+00, 3.67633778e-01, 1.15612262e+00, 2.83469893e+00,
       1.67857630e+00, 4.74152866e+00, 1.10290133e+00, 3.46836787e+00,
       8.50409678e+00, 5.03572890e+00, 2.49699209e+00, 6.50830921e-01,
       1.75422780e+00, 4.93760804e+00, 3.18338025e+00, 1.35920672e+02,
       6.44618194e+01, 7.87595936e+01, 3.75034241e+02, 2.96274647e+02,
       1.58050955e+00, 3.67633778e-01, 1.15612262e+00, 2.83469893e+00,
       1.67857630e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 6.07613078e+00, 1.35970286e+00,
       4.51686635e+00, 1.04388159e+01, 5.92194950e+00, 8.53520510e-02,
       2.90757579e-02, 5.59310778e-02

With this data available for all 243 endmembers, we have an embedding of all these structures in so-called feature space. However, it is so highly dimensional that it is impossible to visualize. Fortunately, we can use dimensionality reduction techniques to reduce the number of dimensions to 2 or 3 and visualize the results for human eyes.

We will use TSNE (t-distributed stochastic neighbor embedding) to reduce the dimensionality to 2 using `sklearn` library. It is not a part of pySIPFENN dependency tree, so you may need to install it with `pip` below (after uncommenting the line).

In [35]:
#!pip install scikit-learn

In [61]:
from sklearn.manifold import TSNE              # neighborhood  dimensionality reduction
from sklearn.preprocessing import MinMaxScaler # scaling

We start by copying the `c.descriptorData`, normalizing it across feature values to minima and maxima using `MinMaxScaler` (similar to fixed Max scaler inside pySIPFENN NNs), and setting up the `TSNE` object. 

In [125]:
scaler = MinMaxScaler()
descriptorData = scaler.fit_transform(np.array(c.descriptorData))

We will use `perplexity=4`, as we don't expect large clusters, `n_iter=1000` to speed up the calculation, `pca` initialization to give a good starting point, and use 
[`scipy.spatial.distance.correlation`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.correlation.html#scipy.spatial.distance.correlation) as the neighbor distance metric between points.

You are encouraged play with these parameters to see how they affect the results!

In [143]:
tsne = TSNE(n_components=2, perplexity=4, init='pca', n_iter=1000, metric='correlation', angle=0.2, n_jobs=-1)

Now, let's embed the data in 2D space and look at the first 3 results as a sanity check.

In [201]:
embeddedDescriptorData = tsne.fit_transform(descriptorData)
pprint(embeddedDescriptorData[:3])

array([[ 33.34489 , -46.64512 ],
       [-17.25321 ,  55.28404 ],
       [ 37.56524 , -43.150795]], dtype=float32)


Note that the embedding is stochastic and chaotic, so you will get different results each time you run it. However, you can easily fix it by setting `random_state` parameter to a fixed value.

In [145]:
tsne.random_state = 580
embeddedDescriptorData = tsne.fit_transform(descriptorData)
pprint(embeddedDescriptorData[:3])

array([[ 33.34489 , -46.64512 ],
       [-17.25321 ,  55.28404 ],
       [ 37.56524 , -43.150795]], dtype=float32)


Finally, we can plot the results using `plotly` library. We will colorize the points by the formation energy coming from the first model in `c.predictions` variable. We will also use `structList` to get chemical formulas when hovering over the points.

In [202]:
fig = px.scatter(x=np.transpose(embeddedDescriptorData)[0],
                 y=np.transpose(embeddedDescriptorData)[1],
                 hover_name=[s.formula for s in structList],
                 color=[round(p[0], 3) for p in c.predictions],
                 color_discrete_sequence=px.colors.qualitative.Dark24,
                 template='plotly_white',
                 labels={'x': f'{len(descriptorData[0])}D->2D TSNE1',
                         'y': f'{len(descriptorData[0])}D->2D TSNE2',
                         'color': f'Formation Energy (eV/atom)'},
                 height=400,
                 width=800
                 )
fig.show()

An immediate result one can see here is that similar structures in the feature space have similar energies in the prediction space, which is a good sign that the method is working as expected. Before moving further, lets upload the results to the database so we can visualize them later in MongoDB Charts if we want to.

In [203]:
for id, embedding in zip(idList, embeddedDescriptorData.tolist()):
    collection.update_one(
        {'_id': id}, 
        {'$set': {'TSNE_2D_X': embedding[0], 'TSNE_2D_Y': embedding[1]}})

In [204]:
collection.find_one({})

{'_id': ObjectId('65300205c113e132c8af4051'),
 'structure': {'@module': 'pymatgen.core.structure',
  '@class': 'Structure',
  'charge': 0,
  'lattice': {'matrix': [[8.547048, 0.0, 0.0],
    [0.0, 8.547048, 0.0],
    [0.0, 0.0, 4.477714]],
   'pbc': [True, True, True],
   'a': 8.547048,
   'b': 8.547048,
   'c': 4.477714,
   'alpha': 90.0,
   'beta': 90.0,
   'gamma': 90.0,
   'volume': 327.10609528461225},
  'properties': {},
  'sites': [{'species': [{'element': 'Fe', 'occu': 1}],
    'abc': [0.737702, 0.063709, 0.0],
    'xyz': [6.305174403696, 0.544523881032, 0.0],
    'properties': {},
    'label': 'Fe'},
   {'species': [{'element': 'Fe', 'occu': 1}],
    'abc': [0.262298, 0.936291, 0.0],
    'xyz': [2.241873596304, 8.002524118968, 0.0],
    'properties': {},
    'label': 'Fe'},
   {'species': [{'element': 'Fe', 'occu': 1}],
    'abc': [0.436291, 0.237702, 0.5],
    'xyz': [3.729000118968, 2.031650403696, 2.238857],
    'properties': {},
    'label': 'Fe'},
   {'species': [{'element

## Transfer Learning on small DFT dataset

Now, let's look at an example of how we can use this data to guide DFT calculations!

There are a few possible ways, but we will focus on transfer learning, which is a technique of using a model trained on a large dataset (e.g. all OQMD, like `NN20` we used, or Materials Project) and fine-tuning it on a smaller dataset (e.g. 10 DFT calcualtions of Sigma phase endmembers). 

In [366]:
dftN = 12

### Selecting a subset of data

***The most critical step to getting good performance here is the selections of a good small subset we will train on.***

This can be done in many ways, but the baseline is a random selection like:

In [367]:
import random

randomEndmembersList = random.sample(range(len(descriptorData)), dftN)
print(randomEndmembersList)

[207, 33, 142, 5, 180, 116, 78, 196, 21, 178, 211, 144]


However, with the feature space embedding, we can *on average* do better than that! Let's look for some representative structures in the feature space by identifying cluster centers with `KMeans` clustering from, again, `sklearn` library, and locating points nearest to them by computing `pairwise_distances` matrix between all points and cluster centers.

In [368]:
from sklearn.cluster import KMeans                          # clustering method
from sklearn.metrics import pairwise_distances_argmin_min   # distance metric

kmeansClustering = KMeans(n_clusters=dftN, n_init=500, max_iter=1000, random_state=580)
clusterData = kmeansClustering.fit(embeddedDescriptorData)
print(clusterData.cluster_centers_)

[[-10.58867   -34.97442  ]
 [ 20.730095   45.967346 ]
 [  9.300632   -1.3748279]
 [-21.391657   48.84668  ]
 [ 50.806084    5.4746685]
 [ 43.81489   -38.45249  ]
 [-47.047497  -33.913918 ]
 [-52.87097    18.821074 ]
 [ 24.432981  -53.157425 ]
 [-10.445501   10.5918665]
 [ 39.026363   31.754313 ]
 [-33.4409    -19.664864 ]]


In [369]:
clusterCenters, _ = pairwise_distances_argmin_min(clusterData.cluster_centers_, embeddedDescriptorData)
print(clusterCenters)

[239 156  63 108 124 216  89  13  26 219  14 169]


which we can now plot on top of the TSNE embedding we made earlier.

In [370]:
fig = px.scatter(x=np.transpose(embeddedDescriptorData)[0],
                 y=np.transpose(embeddedDescriptorData)[1],
                 hover_name=[s.formula for s in structList],
                 color=['cluster center' if i in clusterCenters else 'other' for i in
                        range(len(embeddedDescriptorData))],
                 opacity=0.85,
                 color_discrete_sequence=px.colors.qualitative.Dark24,
                 template='plotly_white',
                 labels={'x': f'{len(descriptorData[0])}D->2D TSNE1',
                         'y': f'{len(descriptorData[0])}D->2D TSNE2',
                         'color': f'Embedded Points'},
                 height=400,
                 width=800
                 )
fig.show()

### Our virtual HPC Run

Before we go forward, in order to prove that a selection of points for DFT was good, we need to calculate all of them, even though only a small subset will be used for training. 

To do so, we will now take results pre-calculated with DFTTK (in `assets/sigma.csv`) and insert them into the database; pretending they were calculated on a HPC cluster. The result will be functionally the same.

Note we will be matching them by permutation (with the same order of elements we used earlier) and not by the `id` as we did earlier as the order of results is not guaranteed to be the same.

In [304]:
with open('assets/sigma.csv', 'r') as sigmaData:
    for l in sigmaData.readlines()[1:]:
        lSplit = l.split(',')
        permutation = "".join(lSplit[0:5])
        DFT_dH = float(lSplit[8])
        collection.update_one({'permutation': permutation}, {'$set': {'DFT_dH': DFT_dH}})

In [305]:
collection.find_one({})

{'_id': ObjectId('65300205c113e132c8af4051'),
 'structure': {'@module': 'pymatgen.core.structure',
  '@class': 'Structure',
  'charge': 0,
  'lattice': {'matrix': [[8.547048, 0.0, 0.0],
    [0.0, 8.547048, 0.0],
    [0.0, 0.0, 4.477714]],
   'pbc': [True, True, True],
   'a': 8.547048,
   'b': 8.547048,
   'c': 4.477714,
   'alpha': 90.0,
   'beta': 90.0,
   'gamma': 90.0,
   'volume': 327.10609528461225},
  'properties': {},
  'sites': [{'species': [{'element': 'Fe', 'occu': 1}],
    'abc': [0.737702, 0.063709, 0.0],
    'xyz': [6.305174403696, 0.544523881032, 0.0],
    'properties': {},
    'label': 'Fe'},
   {'species': [{'element': 'Fe', 'occu': 1}],
    'abc': [0.262298, 0.936291, 0.0],
    'xyz': [2.241873596304, 8.002524118968, 0.0],
    'properties': {},
    'label': 'Fe'},
   {'species': [{'element': 'Fe', 'occu': 1}],
    'abc': [0.436291, 0.237702, 0.5],
    'xyz': [3.729000118968, 2.031650403696, 2.238857],
    'properties': {},
    'label': 'Fe'},
   {'species': [{'element

### Fine-Tuning the Models

Now, we are ready to fine-tune the models! In near fututre release of pySIPFENN, this will be done using a high level API (including hyperparameter tuning), so if you are following this lecture in the future, you may want to check out the documentation for the latest version. For now, we will do it manually by calling a number of `torch` and `onnx` functions, which won't be covered in this lecture.

If you are interested in learning more about what we are doing here, you can have a look at [Section 3.5 in the SIPFENN paper](https://www.sciencedirect.com/science/article/pii/S0927025622000593?via%3Dihub#sec3).

In [501]:
import torch
import onnx
import onnx2torch
import random

Pull DFT data from the database and convert it to a `torch` tensor. The `{'DFT_dH': 1}` after the query is a projection, which means we only want to get the `DFT_dH` field from the database. This is a good practice to reduce the amount of data transferred over the network.

In [477]:
labelTensor = torch.from_numpy(np.array([[collection.find_one({'_id': id}, {'DFT_dH': 1})['DFT_dH']] for id in idList])).float()
print(labelTensor[:10])

tensor([[0.0805],
        [0.0778],
        [0.0802],
        [0.0935],
        [0.0809],
        [0.1118],
        [0.0656],
        [0.0678],
        [0.0702],
        [0.0867]])


Convert the `numpy` array to `torch` tensor.

In [478]:
ddTensor = torch.from_numpy(descriptorData).float()

Load the underlying model, in this case `NN24` as it is the lightest on memory and fasterst to tune, and get it ready.

In [494]:
with resources.files('pysipfenn').joinpath('modelsSIPFENN/SIPFENN_Krajewski2020_NN24.onnx') as nn24model:
    model = onnx2torch.convert(onnx.load(nn24model))
    model.eval()

Set the optimizer and MAE loss function.

In [495]:
optimizer = torch.optim.Adam(model.parameters(), lr=0.00005)
loss = torch.nn.L1Loss()

And before we go further, lets see the current performance of the model on the DFT data we just pulled.

In [496]:
dataOut = model(ddTensor, None)
print(dataOut[:10])

tensor([[ 0.1671],
        [ 0.3394],
        [ 0.1794],
        [ 0.5271],
        [ 0.4225],
        [ 0.4708],
        [ 0.2241],
        [ 0.4732],
        [-0.2594],
        [ 0.4692]], grad_fn=<SliceBackward0>)


In [497]:
loss(dataOut, labelTensor)

tensor(0.2319, grad_fn=<MeanBackward0>)

### Random Selection 

Randomly select the subset of data to train on.

In [498]:
transferIndexes = random.sample(range(len(descriptorData)), dftN)
validationIndexes = list(set(range(len(descriptorData))).difference(transferIndexes))
print(transferIndexes)

[5, 223, 169, 54, 32, 128, 164, 161, 171, 44, 229, 133]


Cherry-pick the data creating new tensors

In [499]:
transferData = torch.index_select(ddTensor, 0, torch.LongTensor(transferIndexes)).float()
transferLabels = torch.index_select(labelTensor, 0, torch.LongTensor(transferIndexes)).float()

validationData = torch.index_select(ddTensor, 0, torch.LongTensor(validationIndexes)).float()
validationLabels = torch.index_select(labelTensor, 0, torch.LongTensor(validationIndexes)).float()

***And finally, train the model!***

In [500]:
model.eval()
transferLosses = [float(loss(model(transferData, None), transferLabels))]
validationLosses = [float(loss(model(validationData, None), validationLabels))]

model.train()
print('Initial Mean Absolute Errors (MAE):')
print(f'Train: {round(1000 * float(transferLosses[0]), 1):>5}  |  Val: {round(1000 * float(validationLosses[0]), 1):>5}   [meV/atom]')
print('Starting Training...')
for i in range(250):
    optimizer.zero_grad()
    transfer_pred = model(transferData, None)
    transfer_loss = loss(transfer_pred, transferLabels)
    transferLosses.append(float(transfer_loss))
    transfer_loss.backward()
    optimizer.step()
    validationLosses.append(float(loss(model(validationData, None), validationLabels)))
    if i % 10 == 0:
        print(f'Epoch {i:>3}:  Train: {round(1000 * float(transferLosses[-1]), 1):>5}  |  Val: {round(1000 * float(validationLosses[-1]), 1):>5}   [meV/atom]')
print('Training Complete!')
model.eval()
transferLosses.append(float(loss(model(transferData, None), transferLabels)))
validationLosses.append(float(loss(model(validationData, None), validationLabels)))
print('Final Evaluation Mean Absolute Error (MAE):')
print(f'Train: {round(1000 * float(transferLosses[-1]), 1):>5}  |  Val: {round(1000 * float(validationLosses[-1]), 1):>5}')

Initial Mean Absolute Errors (MAE):
Train: 234.1  |  Val: 231.8   [meV/atom]
Starting Training...
Epoch   0:  Train: 255.0  |  Val: 203.1   [meV/atom]
Epoch  10:  Train: 141.3  |  Val: 117.2   [meV/atom]
Epoch  20:  Train:  28.1  |  Val:  79.9   [meV/atom]
Epoch  30:  Train:  38.2  |  Val:  69.0   [meV/atom]
Epoch  40:  Train:  20.0  |  Val:  56.5   [meV/atom]
Epoch  50:  Train:  24.6  |  Val:  57.9   [meV/atom]
Epoch  60:  Train:  24.2  |  Val:  53.3   [meV/atom]
Epoch  70:  Train:  21.4  |  Val:  51.8   [meV/atom]
Epoch  80:  Train:  25.6  |  Val:  52.9   [meV/atom]
Epoch  90:  Train:  29.4  |  Val:  49.2   [meV/atom]
Epoch 100:  Train:  23.2  |  Val:  44.1   [meV/atom]
Epoch 110:  Train:  12.6  |  Val:  44.6   [meV/atom]
Epoch 120:  Train:   8.7  |  Val:  42.6   [meV/atom]
Epoch 130:  Train:  19.6  |  Val:  40.7   [meV/atom]
Epoch 140:  Train:  20.2  |  Val:  40.6   [meV/atom]
Epoch 150:  Train:  19.4  |  Val:  40.7   [meV/atom]
Epoch 160:  Train:  18.4  |  Val:  39.9   [meV/atom]
E

### Feature-Space-Informed Selection

Now, lets do the same, but using the subset of data we selected based on the feature space embedding and assigned to `clusterCenters` variable.

Start by reloading feature data from pySIPFENN.

In [486]:
with resources.files('pysipfenn').joinpath('modelsSIPFENN/SIPFENN_Krajewski2020_NN24.onnx') as nn24model:
    model = onnx2torch.convert(onnx.load(nn24model))
    model.eval()

optimizer = torch.optim.Adam(model.parameters(), lr=0.00005)
loss = torch.nn.L1Loss()

Select the subset with `clusterCenters` and convert it to `torch` tensor.

In [487]:
transferIndexes = clusterCenters
validationIndexes = list(set(range(len(descriptorData))).difference(clusterCenters))
print(transferIndexes)

[239 156  63 108 124 216  89  13  26 219  14 169]


In [488]:
transferData = torch.index_select(ddTensor, 0, torch.LongTensor(transferIndexes)).float()
transferLabels = torch.index_select(labelTensor, 0, torch.LongTensor(transferIndexes)).float()

validationData = torch.index_select(ddTensor, 0, torch.LongTensor(validationIndexes)).float()
validationLabels = torch.index_select(labelTensor, 0, torch.LongTensor(validationIndexes)).float()

In [489]:
model.eval()
transferLosses = [float(loss(model(transferData, None), transferLabels))]
validationLosses = [float(loss(model(validationData, None), validationLabels))]

model.train()
print('Initial Mean Absolute Errors (MAE):')
print(f'Train: {round(1000 * float(transferLosses[0]), 1):>5}  |  Val: {round(1000 * float(validationLosses[0]), 1):>5}   [meV/atom]')
print('Starting Training...')
for i in range(250):
    optimizer.zero_grad()
    transfer_pred = model(transferData, None)
    transfer_loss = loss(transfer_pred, transferLabels)
    transferLosses.append(float(transfer_loss))
    transfer_loss.backward()
    optimizer.step()
    validationLosses.append(float(loss(model(validationData, None), validationLabels)))
    if i % 10 == 0:
        print(f'Epoch {i:>3}:  Train: {round(1000 * float(transferLosses[-1]), 1):>5}  |  Val: {round(1000 * float(validationLosses[-1]), 1):>5}')
print('Training Complete!')
model.eval()
transferLosses.append(float(loss(model(transferData, None), transferLabels)))
validationLosses.append(float(loss(model(validationData, None), validationLabels)))
print('Final Evaluation Mean Absolute Error (MAE):')
print(f'Train: {round(1000 * float(transferLosses[-1]), 1):>5}  |  Val: {round(1000 * float(validationLosses[-1]), 1):>5}')

Initial Mean Absolute Errors (MAE):
Train: 268.0  |  Val: 230.0   [meV/atom]
Starting Training...
Epoch   0:  Train: 268.8  |  Val: 205.5
Epoch  10:  Train:  91.9  |  Val:  97.7
Epoch  20:  Train:  56.7  |  Val:  69.6
Epoch  30:  Train:  31.6  |  Val:  58.1
Epoch  40:  Train:  45.0  |  Val:  54.3
Epoch  50:  Train:  32.6  |  Val:  52.8
Epoch  60:  Train:  23.2  |  Val:  44.9
Epoch  70:  Train:  20.5  |  Val:  46.8
Epoch  80:  Train:  14.1  |  Val:  44.3
Epoch  90:  Train:  25.5  |  Val:  40.7
Epoch 100:  Train:  25.4  |  Val:  41.6
Epoch 110:  Train:  14.8  |  Val:  41.6
Epoch 120:  Train:  17.4  |  Val:  38.5
Epoch 130:  Train:  18.7  |  Val:  36.0
Epoch 140:  Train:  21.9  |  Val:  35.2
Epoch 150:  Train:  14.3  |  Val:  34.7
Epoch 160:  Train:  18.2  |  Val:  34.4
Epoch 170:  Train:  17.5  |  Val:  34.0
Epoch 180:  Train:  16.1  |  Val:  30.8
Epoch 190:  Train:  17.4  |  Val:  33.7
Epoch 200:  Train:  24.9  |  Val:  30.1
Epoch 210:  Train:  16.7  |  Val:  32.6
Epoch 220:  Train:  17

Now, **on average you should see a result around 25meV/atom** depending on the run, which should be around 20-30% reduction in MAE, which is a significant improvement! However, the real power of this approach lays in its consistency, as if you repeat this process many times, the random selection may sometimes happen to be better, but it will ocassionally be two or three times higher.

## Conclusions and Further Resources

If you were able to complete this notebook, you should now have a good basic understanding of how manipulte atomic configurations in Python, send it back and forth to MongoDB, and use pySIPFENN to (1) predict formation energy, (2) featurize structures, and (3) tune ML models.

Here are some additional resources you may find useful if you want to learn more:

- [MongoDB Compass GUI Application](https://www.mongodb.com/try/download/compass) which will allow you to interact with the database in a very user-friendly way, including basic visualization of the data, testing your queries, analyzing the structure of the database, and more.
- [MongoDB Query Language (MQL) documentation](https://docs.mongodb.com/manual/tutorial/query-documents/)
- [pySIPFENN documentation](https://pysipfenn.org) and [pySIPFENN GitHub repository](https://git.pysipfenn.org)
- [Results section of SIPFENN Paper](https://www.sciencedirect.com/science/article/pii/S0927025622000593?via%3Dihub#sec3) for a discussion of how and why we do transfer learning. Including reasoning behing some hyperparameters we used here.